File: src/incode_mcp/staging.py
"""Journalled Arrow staging for crash-recoverable index commits.

An index run never touches the live Lance tables while it scans, parses, and
embeds. It streams file and chunk rows into Arrow IPC files under
``<data>/staging/<project-id>/<job-id>/`` and records its progress in
``journal.json``. Only once staging finishes does the run record the live
tables' versions, switch the journal to ``committing``, and apply the staged
batches. A crash before that switch leaves the live tables untouched; a crash
after it is rolled back to the recorded versions on the next startup, so the
previously searchable generation of the project survives either way.

Every file write goes through a temporary sibling, an ``fsync``, and an atomic
rename, so a journal or Arrow payload is never observed half-written.
"""

from __future__ import annotations

import json
import logging
import os
import shutil
import time
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path
from typing import Any, cast

import numpy as np
import pyarrow as pa

from .models import StoredFile
from .storage import LanceStore, TableVersions

logger = logging.getLogger(__name__)

JOURNAL_NAME = "journal.json"
FILES_NAME = "files.arrow"
CHUNKS_NAME = "chunks.arrow"

PHASE_STAGING = "staging"
PHASE_COMMITTING = "committing"
PHASE_COMPLETE = "complete"
PHASE_ROLLED_BACK = "rolled_back"

JOURNAL_FORMAT_VERSION = 1


@dataclass(frozen=True)
class ChunkRow:
    """One staged chunk: the StoredChunk fields, with the vector still packed.

    ``vector`` is contiguous little-endian float32 bytes exactly as the
    embedding worker returned them, so the write path never materializes a
    list of Python floats per chunk.
    """

    chunk_id: str
    file_id: str
    project_id: str
    path: str
    language: str
    kind: str
    symbol: str | None
    qualified_symbol: str | None
    parent_symbol: str | None
    start_byte: int
    end_byte: int
    start_line: int
    end_line: int
    content: str
    embedding_text: str
    search_text: str
    content_hash: str
    part_index: int
    vector: bytes


def _write_atomically(path: Path, payload: bytes) -> None:
    """Write *payload* to *path* through a temporary sibling and fsync."""
    temporary = path.with_name(f"{path.name}.tmp")
    with temporary.open("wb") as sink:
        sink.write(payload)
        sink.flush()
        os.fsync(sink.fileno())
    os.replace(temporary, path)
    _sync_directory(path.parent)


def _sync_directory(directory: Path) -> None:
    """Best-effort fsync of a directory so renames inside it are durable."""
    if not hasattr(os, "O_DIRECTORY"):
        return
    try:
        descriptor = os.open(directory, os.O_RDONLY | os.O_DIRECTORY)
    except OSError:
        return
    try:
        os.fsync(descriptor)
    except OSError:
        pass
    finally:
        os.close(descriptor)


class StagingJob:
    """One index run's staged output and its journal.

    Writers open in :meth:`begin` and stay open while the run stages rows;
    :meth:`begin_commit` finalizes the Arrow payloads and records the live
    table versions the commit must be able to roll back to. Batches are
    written per embedding group of a single file, so a record batch never
    spans two files -- the commit can stream them back one file at a time.
    """

    def __init__(
        self,
        staging_root: Path,
        project_id: str,
        *,
        file_schema: pa.Schema,
        chunk_schema: pa.Schema,
        job_id: str | None = None,
    ) -> None:
        self.staging_root = staging_root
        self.project_id = project_id
        self.job_id = job_id or f"{time.time_ns():x}-{os.getpid()}"
        self._file_schema = file_schema
        self._chunk_schema = chunk_schema
        self.replace_file_ids: list[str] = []
        self.removed_file_ids: list[str] = []
        self._journal: dict[str, Any] = {}
        self._files_sink: Any = None
        self._chunks_sink: Any = None
        self._files_writer: pa.RecordBatchWriter | None = None
        self._chunks_writer: pa.RecordBatchWriter | None = None

    @property
    def directory(self) -> Path:
        return self.staging_root / self.project_id / self.job_id

    def begin(self) -> None:
        self.directory.mkdir(parents=True, exist_ok=True)
        self._journal = {
            "version": JOURNAL_FORMAT_VERSION,
            "job_id": self.job_id,
            "project_id": self.project_id,
            "phase": PHASE_STAGING,
            "created_at_ns": time.time_ns(),
        }
        self._write_journal()
        self._files_sink, self._files_writer = self._open_writer(FILES_NAME, self._file_schema)
        self._chunks_sink, self._chunks_writer = self._open_writer(CHUNKS_NAME, self._chunk_schema)

    def stage_file(self, record: StoredFile) -> None:
        """Stage one file record. Files are few and carry no vectors."""
        assert self._files_writer is not None
        batch = pa.RecordBatch.from_pylist([record.model_dump()], schema=self._file_schema)
        self._files_writer.write_batch(batch)

    def stage_chunks(self, rows: list[ChunkRow]) -> None:
        """Stage one embedding group's chunk rows as a single record batch."""
        assert self._chunks_writer is not None
        if not rows:
            return
        vector_type = self._chunk_schema.field("vector").type
        dimension = vector_type.list_size
        columns: list[pa.Array] = []
        for field in self._chunk_schema:
            if field.name == "vector":
                packed = b"".join(row.vector for row in rows)
                flat = np.frombuffer(packed, dtype="<f4")
                values = pa.array(flat, type=pa.float32())
                columns.append(pa.FixedSizeListArray.from_arrays(values, dimension))
            else:
                columns.append(
                    pa.array([getattr(row, field.name) for row in rows], type=field.type)
                )
        self._chunks_writer.write_batch(pa.record_batch(columns, schema=self._chunk_schema))

    def mark_replaced(self, file_id: str) -> None:
        self.replace_file_ids.append(file_id)

    def mark_removed(self, file_id: str) -> None:
        self.removed_file_ids.append(file_id)

    def begin_commit(self, versions: TableVersions) -> None:
        """Finalize the payloads and record the versions a rollback restores."""
        self._close_writers()
        self._journal.update(
            {
                "phase": PHASE_COMMITTING,
                "files_version": versions.files,
                "chunks_version": versions.chunks,
                "replace_file_ids": self.replace_file_ids,
                "removed_file_ids": self.removed_file_ids,
            }
        )
        self._write_journal()

    def files_table(self) -> pa.Table:
        """Read back every staged file record. One row per file; small."""
        return pa.ipc.open_file(self.directory / FILES_NAME).read_all()

    def iter_chunk_groups(self) -> Iterator[tuple[str, pa.Table]]:
        """Yield ``(file_id, table)`` for each replaced file, one at a time.

        Staged rows for files absent from ``replace_file_ids`` -- files that
        failed mid-run -- are skipped, and a replaced file with no staged rows
        yields an empty table so the commit deletes its previous chunks.
        """
        wanted = set(self.replace_file_ids)
        seen: set[str] = set()
        current_file_id: str | None = None
        batches: list[pa.RecordBatch] = []
        reader = pa.ipc.open_file(self.directory / CHUNKS_NAME)
        for index in range(reader.num_record_batches):
            batch = reader.get_batch(index)
            file_id = cast(str, batch.column("file_id")[0].as_py())
            if file_id != current_file_id:
                if batches:
                    assert current_file_id is not None
                    seen.add(current_file_id)
                    yield current_file_id, pa.Table.from_batches(batches, schema=self._chunk_schema)
                    batches = []
                current_file_id = file_id
            if file_id in wanted:
                batches.append(batch)
        if batches:
            assert current_file_id is not None
            seen.add(current_file_id)
            yield current_file_id, pa.Table.from_batches(batches, schema=self._chunk_schema)
        empty = pa.Table.from_batches([], schema=self._chunk_schema)
        for file_id in self.replace_file_ids:
            if file_id not in seen:
                yield file_id, empty

    def complete(self) -> None:
        """Mark the commit successful and remove the staged directory."""
        self._journal["phase"] = PHASE_COMPLETE
        self._write_journal()
        shutil.rmtree(self.directory)

    def rolled_back(self) -> None:
        """Mark the commit rolled back and remove the staged directory."""
        self._journal["phase"] = PHASE_ROLLED_BACK
        self._write_journal()
        shutil.rmtree(self.directory)

    def discard(self) -> None:
        """Abandon a job that never started committing; live tables untouched."""
        try:
            self._close_writers(finalize=False)
        except Exception:
            logger.debug("Discarding staging writers failed", exc_info=True)
        shutil.rmtree(self.directory, ignore_errors=True)

    def _open_writer(self, name: str, schema: pa.Schema) -> tuple[Any, pa.RecordBatchWriter]:
        temporary = self.directory / f"{name}.tmp"
        sink = temporary.open("wb")
        return sink, pa.ipc.new_file(sink, schema)

    def _close_writers(self, *, finalize: bool = True) -> None:
        for name, sink, writer in (
            (FILES_NAME, self._files_sink, self._files_writer),
            (CHUNKS_NAME, self._chunks_sink, self._chunks_writer),
        ):
            if writer is None:
                continue
            writer.close()
            sink.flush()
            os.fsync(sink.fileno())
            sink.close()
            if finalize:
                os.replace(
                    self.directory / f"{name}.tmp",
                    self.directory / name,
                )
        self._files_sink = self._chunks_sink = None
        self._files_writer = self._chunks_writer = None
        if finalize:
            _sync_directory(self.directory)

    def _write_journal(self) -> None:
        payload = json.dumps(self._journal, indent=2, sort_keys=True).encode()
        _write_atomically(self.directory / JOURNAL_NAME, payload)


def recover_staged_commits(staging_root: Path, store: LanceStore) -> int:
    """Roll back every interrupted commit before queries are accepted.

    A journal still in ``staging`` never reached the live tables, and one in a
    terminal phase only failed to clean up, so both are simply removed. A
    journal in ``committing`` means a crash after live writes began: restore
    both tables to the recorded versions, mark the journal ``rolled_back``,
    and remove the directory. Restoring an already-restored table is safe, so
    repeated recovery over the same journal is idempotent.
    """
    if not staging_root.is_dir():
        return 0
    recovered = 0
    for journal_path in sorted(staging_root.glob(f"*/*/{JOURNAL_NAME}")):
        directory = journal_path.parent
        try:
            journal = json.loads(journal_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            logger.warning("Ignoring unreadable staging journal: %s", journal_path)
            continue
        if journal.get("phase") != PHASE_COMMITTING:
            shutil.rmtree(directory, ignore_errors=True)
            continue
        try:
            store.restore_versions(
                str(journal["project_id"]),
                TableVersions(
                    files=int(journal["files_version"]),
                    chunks=int(journal["chunks_version"]),
                ),
            )
        except Exception:
            # Leave the journal in place so the next startup retries instead of
            # silently serving a half-committed project.
            logger.exception("Could not roll back interrupted commit in %s", directory)
            continue
        journal["phase"] = PHASE_ROLLED_BACK
        _write_atomically(journal_path, json.dumps(journal, indent=2, sort_keys=True).encode())
        shutil.rmtree(directory, ignore_errors=True)
        recovered += 1
    return recovered

File: ts/packages/server/src/staging.ts
/**
 * Journalled Arrow staging for crash-recoverable index commits.
 *
 * A run writes its file, chunk, and reference rows beneath
 * `<data>/staging/<project-id>/<job-id>/` before mutating Lance.  The journal
 * moves to `committing` only after the Arrow IPC payloads are durable and the
 * live table versions have been recorded, leaving startup recovery enough
 * information to restore a partially applied commit.
 */

import { randomUUID } from "node:crypto";
import type { Dirent, WriteStream } from "node:fs";
import { createWriteStream } from "node:fs";
import fs from "node:fs/promises";
import path from "node:path";
import { finished } from "node:stream/promises";
import {
  makeData,
  Precision,
  RecordBatch,
  RecordBatchFileWriter,
  RecordBatchReader,
  type Schema,
  Struct,
  Table,
  tableToIPC,
  vectorFromArray,
} from "apache-arrow";
import type { StoredChunk, StoredFile } from "./models.ts";
import type { ReferenceRecord } from "./reference-store.ts";
import type { LanceStore, ReplacementBatch, TableVersions } from "./storage.ts";

export const JOURNAL_NAME = "journal.json";
export const FILES_NAME = "files.arrow";
export const CHUNKS_NAME = "chunks.arrow";
export const REFERENCES_NAME = "references.arrow";

export const PHASE_STAGING = "staging";
export const PHASE_COMMITTING = "committing";
export const PHASE_COMPLETE = "complete";
export const PHASE_ROLLED_BACK = "rolled_back";

export const LEGACY_JOURNAL_FORMAT_VERSION = 1;
export const JOURNAL_FORMAT_VERSION = 2;
export const MAX_RECOVERY_ATTEMPTS = 3;

export const COMMIT_BATCH_MAX_FILES = 64;
export const COMMIT_BATCH_MAX_ROWS = 20_000;
export const COMMIT_BATCH_MAX_BYTES = 64 * 1024 * 1024;

/** The TypeScript model keeps vectors unpacked; Arrow owns their IPC encoding. */
export type ChunkRow = StoredChunk;
export type ReferenceRow = ReferenceRecord;

type JournalPhase =
  | typeof PHASE_STAGING
  | typeof PHASE_COMMITTING
  | typeof PHASE_COMPLETE
  | typeof PHASE_ROLLED_BACK;

interface StagingJournal {
  version: number;
  job_id: string;
  project_id: string;
  phase: JournalPhase | string;
  created_at_ns: number;
  files_version?: number;
  chunks_version?: number;
  references_version?: number;
  replace_file_ids?: string[];
  replace_reference_file_ids?: string[];
  removed_file_ids?: string[];
  recovery_attempts?: number;
}

interface PayloadWriter {
  schema: Schema;
  writer: RecordBatchFileWriter;
  output: WriteStream;
  temporary: string;
  wrote: boolean;
}

export type StagingStore = Pick<
  LanceStore,
  "restoreVersions" | "markProjectState" | "replaceFilesFromArrow" | "tableVersions"
>;

export interface StagingJobOptions {
  fileSchema: Schema;
  chunkSchema: Schema;
  referenceSchema: Schema;
  jobId?: string;
}

/** Write a whole small control file, then atomically publish and sync it. */
async function writeAtomically(target: string, payload: string): Promise<void> {
  const temporary = `${target}.tmp`;
  let handle: fs.FileHandle | null = null;
  try {
    handle = await fs.open(temporary, "wx");
    await handle.writeFile(payload, "utf8");
    await handle.sync();
    await handle.close();
    handle = null;
    await fs.rename(temporary, target);
    await syncDirectory(path.dirname(target));
  } catch (error) {
    if (handle !== null) await handle.close().catch(() => undefined);
    await fs.rm(temporary, { force: true }).catch(() => undefined);
    throw error;
  }
}

/** Directory fsync is unavailable on Windows and some filesystems. */
async function syncDirectory(directory: string): Promise<void> {
  let handle: fs.FileHandle | null = null;
  try {
    handle = await fs.open(directory, "r");
    await handle.sync();
  } catch {
    // The preceding file fsync and rename still provide the strongest guarantee
    // supported by platforms that cannot open a directory as a file.
  } finally {
    await handle?.close().catch(() => undefined);
  }
}

/** Whether maintenance must retain table versions for a possibly interrupted commit. */
export async function hasPendingRecovery(stagingRoot: string, projectId: string): Promise<boolean> {
  const projectDirectory = path.join(stagingRoot, projectId);
  let jobs: Dirent[];
  try {
    jobs = await fs.readdir(projectDirectory, { withFileTypes: true });
  } catch (error) {
    return !isMissing(error);
  }
  const known = new Set<string>([
    PHASE_STAGING,
    PHASE_COMMITTING,
    PHASE_COMPLETE,
    PHASE_ROLLED_BACK,
  ]);
  for (const job of jobs) {
    if (!job.isDirectory()) continue;
    try {
      const journal = parseJournal(
        await fs.readFile(path.join(projectDirectory, job.name, JOURNAL_NAME), "utf8"),
      );
      if (journal.phase === PHASE_COMMITTING || !known.has(journal.phase)) return true;
    } catch (error) {
      if (isMissing(error)) continue;
      return true;
    }
  }
  return false;
}

/** One index run's staged Arrow payloads and recovery journal. */
export class StagingJob {
  readonly stagingRoot: string;
  readonly projectId: string;
  readonly jobId: string;
  readonly replaceFileIds: string[] = [];
  readonly replaceReferenceFileIds: string[] = [];
  readonly removedFileIds: string[] = [];

  #journal: StagingJournal | null = null;
  #files: PayloadWriter | null = null;
  #chunks: PayloadWriter | null = null;
  #references: PayloadWriter | null = null;
  #chunkFileIds = new Set<string>();
  #referenceFileIds = new Set<string>();
  #lastChunkFileId: string | null = null;
  #lastReferenceFileId: string | null = null;
  #fileSchema: Schema;
  #chunkSchema: Schema;
  #referenceSchema: Schema;

  constructor(stagingRoot: string, projectId: string, options: StagingJobOptions) {
    this.stagingRoot = stagingRoot;
    this.projectId = projectId;
    this.jobId = options.jobId ?? `${Date.now().toString(16)}-${randomUUID()}`;
    this.#fileSchema = options.fileSchema;
    this.#chunkSchema = options.chunkSchema;
    this.#referenceSchema = options.referenceSchema;
  }

  get directory(): string {
    return path.join(this.stagingRoot, this.projectId, this.jobId);
  }

  async begin(): Promise<void> {
    if (this.#journal !== null) throw new Error("staging job has already begun");
    await fs.mkdir(this.directory, { recursive: true });
    this.#journal = {
      version: JOURNAL_FORMAT_VERSION,
      job_id: this.jobId,
      project_id: this.projectId,
      phase: PHASE_STAGING,
      created_at_ns: Date.now() * 1_000_000,
    };
    await this.#writeJournal();
    this.#files = this.#openWriter(FILES_NAME, this.#fileSchema);
    this.#chunks = this.#openWriter(CHUNKS_NAME, this.#chunkSchema);
    this.#references = this.#openWriter(REFERENCES_NAME, this.#referenceSchema);
  }

  async stageFile(record: StoredFile): Promise<void> {
    await this.#writeRows(this.#requireWriter(this.#files), [record]);
  }

  async stageChunks(rows: readonly ChunkRow[]): Promise<void> {
    const first = rows[0];
    if (first === undefined) return;
    const fileId = first.file_id;
    if (rows.some((row) => row.file_id !== fileId)) {
      throw new Error("a staged chunk batch must contain rows from one file");
    }
    if (this.#chunkFileIds.has(fileId) && this.#lastChunkFileId !== fileId) {
      throw new Error("staged chunk batches for a file must be contiguous");
    }
    this.#chunkFileIds.add(fileId);
    this.#lastChunkFileId = fileId;
    await this.#writeRows(this.#requireWriter(this.#chunks), rows);
  }

  async stageReferences(rows: readonly ReferenceRow[]): Promise<void> {
    const first = rows[0];
    if (first === undefined) return;
    const fileId = first.file_id;
    if (rows.some((row) => row.file_id !== fileId)) {
      throw new Error("a staged reference batch must contain rows from one file");
    }
    if (this.#referenceFileIds.has(fileId) && this.#lastReferenceFileId !== fileId) {
      throw new Error("staged reference batches for a file must be contiguous");
    }
    this.#referenceFileIds.add(fileId);
    this.#lastReferenceFileId = fileId;
    await this.#writeRows(this.#requireWriter(this.#references), rows);
  }

  markReplaced(fileId: string): void {
    addOnce(this.replaceFileIds, fileId);
  }

  markReferencesReplaced(fileId: string): void {
    addOnce(this.replaceReferenceFileIds, fileId);
  }

  markRemoved(fileId: string): void {
    addOnce(this.removedFileIds, fileId);
  }

  async beginCommit(versions: TableVersions): Promise<void> {
    this.#requireJournal();
    await this.#closeWriters();
    this.#journal = {
      ...this.#requireJournal(),
      phase: PHASE_COMMITTING,
      files_version: versions.files,
      chunks_version: versions.chunks,
      references_version: versions.references,
      replace_file_ids: [...this.replaceFileIds],
      replace_reference_file_ids: [...this.replaceReferenceFileIds],
      removed_file_ids: [...this.removedFileIds],
    };
    await this.#writeJournal();
  }

  /** The storage API accepts decoded rows; the durable source is Arrow IPC. */
  async filesTable(): Promise<StoredFile[]> {
    return this.#readRows<StoredFile>(FILES_NAME);
  }

  async iterChunkBatches(options: BatchOptions = {}): Promise<ReplacementBatch<StoredChunk>[]> {
    return this.#batches<StoredChunk>(CHUNKS_NAME, this.replaceFileIds, options);
  }

  async iterReferenceBatches(
    options: BatchOptions = {},
  ): Promise<ReplacementBatch<ReferenceRecord>[]> {
    return this.#batches<ReferenceRecord>(REFERENCES_NAME, this.replaceReferenceFileIds, options);
  }

  async complete(): Promise<void> {
    this.#setTerminalPhase(PHASE_COMPLETE);
    await this.#writeJournal();
    await fs.rm(this.directory, { recursive: true, force: true });
  }

  async rolledBack(): Promise<void> {
    this.#setTerminalPhase(PHASE_ROLLED_BACK);
    await this.#writeJournal();
    await fs.rm(this.directory, { recursive: true, force: true });
  }

  async discard(): Promise<void> {
    if (this.#journal?.phase === PHASE_COMMITTING) return;
    await this.#closeWriters(false).catch(() => undefined);
    await fs.rm(this.directory, { recursive: true, force: true });
  }

  #openWriter(name: string, schema: Schema): PayloadWriter {
    const temporary = path.join(this.directory, `${name}.tmp`);
    const output = createWriteStream(temporary, { flags: "wx" });
    const writer = new RecordBatchFileWriter();
    writer.toNodeStream().pipe(output);
    return { schema, writer, output, temporary, wrote: false };
  }

  async #writeRows<T extends object>(payload: PayloadWriter, rows: readonly T[]): Promise<void> {
    // `RecordBatchFileWriter.write(Table)` finishes an auto-destroy writer.
    // Feed its batches directly so one file writer remains open for the run.
    for (const batch of tableFromRows(payload.schema, rows).batches) payload.writer.write(batch);
    payload.wrote = true;
  }

  async #closeWriters(finalize = true): Promise<void> {
    const writers = [this.#files, this.#chunks, this.#references].filter(
      (writer): writer is PayloadWriter => writer !== null,
    );
    this.#files = this.#chunks = this.#references = null;
    for (const payload of writers) {
      if (finalize && !payload.wrote) await this.#writeRows(payload, []);
      if (finalize) payload.writer.close();
      else payload.writer.abort();
      if (finalize) {
        await finished(payload.output);
        const handle = await fs.open(payload.temporary, "r");
        try {
          await handle.sync();
        } finally {
          await handle.close();
        }
        await fs.rename(payload.temporary, payload.temporary.slice(0, -4));
      } else {
        payload.output.destroy();
      }
    }
    if (finalize) await syncDirectory(this.directory);
  }

  async #readRows<T>(name: string): Promise<T[]> {
    const rows: T[] = [];
    for await (const batch of this.#recordBatches(name)) {
      rows.push(...batch.toArray().map((row) => normalizeRow(row, batch.schema) as T));
    }
    return rows;
  }

  async *#recordBatches(name: string): AsyncGenerator<RecordBatch> {
    const handle = await fs.open(path.join(this.directory, name), "r");
    try {
      const reader = await RecordBatchReader.from(handle);
      for await (const batch of reader) yield batch;
    } finally {
      await handle.close().catch(() => undefined);
    }
  }

  async #batches<T extends { file_id: string }>(
    name: string,
    wantedIds: readonly string[],
    options: BatchOptions,
  ): Promise<ReplacementBatch<T>[]> {
    const maxFiles = options.maxFiles ?? COMMIT_BATCH_MAX_FILES;
    const maxRows = options.maxRows ?? COMMIT_BATCH_MAX_ROWS;
    const maxBytes = options.maxBytes ?? COMMIT_BATCH_MAX_BYTES;
    for (const [name2, value] of Object.entries({ maxFiles, maxRows, maxBytes })) {
      if (!Number.isInteger(value) || value < 1)
        throw new Error(`${name2} must be a positive integer`);
    }
    const wanted = new Set(wantedIds);
    const seen = new Set<string>();
    const groups: Array<{ fileId: string; rows: T[] }> = [];
    let currentFileId: string | null = null;
    let currentRows: T[] = [];
    for await (const batch of this.#recordBatches(name)) {
      for (const item of batch.toArray()) {
        const row = normalizeRow(item, batch.schema) as T;
        if (row.file_id !== currentFileId) {
          if (currentFileId !== null && wanted.has(currentFileId)) {
            groups.push({ fileId: currentFileId, rows: currentRows });
          }
          currentFileId = row.file_id;
          currentRows = [];
        }
        currentRows.push(row);
      }
    }
    if (currentFileId !== null && wanted.has(currentFileId)) {
      groups.push({ fileId: currentFileId, rows: currentRows });
    }
    for (const group of groups) seen.add(group.fileId);
    for (const fileId of wantedIds) {
      if (!seen.has(fileId)) groups.push({ fileId, rows: [] });
    }

    const result: ReplacementBatch<T>[] = [];
    let fileIds: string[] = [];
    let rows: T[] = [];
    let bytes = 0;
    const release = (): void => {
      if (fileIds.length > 0) result.push({ fileIds, rows });
      fileIds = [];
      rows = [];
      bytes = 0;
    };
    for (const group of groups) {
      const groupBytes =
        group.rows.length === 0
          ? 0
          : tableToBytes(
              name === CHUNKS_NAME ? this.#chunkSchema : this.#referenceSchema,
              group.rows,
            );
      if (
        fileIds.length > 0 &&
        (fileIds.length >= maxFiles ||
          rows.length + group.rows.length > maxRows ||
          bytes + groupBytes > maxBytes)
      ) {
        release();
      }
      fileIds.push(group.fileId);
      rows.push(...group.rows);
      bytes += groupBytes;
    }
    release();
    return result;
  }

  #requireWriter(writer: PayloadWriter | null): PayloadWriter {
    if (writer === null) throw new Error("staging payload is not open");
    return writer;
  }

  #requireJournal(): StagingJournal {
    if (this.#journal === null) throw new Error("staging job has not begun");
    return this.#journal;
  }

  #setTerminalPhase(phase: typeof PHASE_COMPLETE | typeof PHASE_ROLLED_BACK): void {
    this.#journal = { ...this.#requireJournal(), phase };
  }

  async #writeJournal(): Promise<void> {
    await writeAtomically(
      path.join(this.directory, JOURNAL_NAME),
      JSON.stringify(this.#requireJournal()),
    );
  }
}

export interface BatchOptions {
  maxFiles?: number;
  maxRows?: number;
  maxBytes?: number;
}

/** Roll back every interrupted commit before the store accepts queries. */
export async function recoverStagedCommits(
  stagingRoot: string,
  store: StagingStore,
): Promise<number> {
  let projectEntries: Dirent[];
  try {
    projectEntries = await fs.readdir(stagingRoot, { withFileTypes: true });
  } catch (error) {
    if (isMissing(error)) return 0;
    throw error;
  }
  let recovered = 0;
  for (const project of projectEntries.sort((left, right) => left.name.localeCompare(right.name))) {
    if (!project.isDirectory()) continue;
    const projectDirectory = path.join(stagingRoot, project.name);
    const jobs = await fs
      .readdir(projectDirectory, { withFileTypes: true })
      .catch(() => [] as Dirent[]);
    for (const job of jobs.sort((left, right) => left.name.localeCompare(right.name))) {
      if (!job.isDirectory()) continue;
      const directory = path.join(projectDirectory, job.name);
      const journalPath = path.join(directory, JOURNAL_NAME);
      let journal: StagingJournal;
      try {
        journal = parseJournal(await fs.readFile(journalPath, "utf8"));
      } catch {
        // An unreadable journal may be repaired manually; removing it could
        // discard the only record of the versions a crash recovery needs.
        continue;
      }
      if (journal.phase !== PHASE_COMMITTING) {
        await fs.rm(directory, { recursive: true, force: true });
        continue;
      }
      try {
        const version = Number(journal.version ?? LEGACY_JOURNAL_FORMAT_VERSION);
        const legacy = version === LEGACY_JOURNAL_FORMAT_VERSION;
        if (!legacy && version !== JOURNAL_FORMAT_VERSION) {
          throw new Error(`unsupported staging journal version: ${version}`);
        }
        const versions = {
          files: journalNumber(journal, "files_version"),
          chunks: journalNumber(journal, "chunks_version"),
          references: legacy ? 0 : journalNumber(journal, "references_version"),
        };
        const restored = await store.restoreVersions(journal.project_id, versions, {
          restoreReferences: !legacy,
        });
        if (!restored) {
          await fs.rm(directory, { recursive: true, force: true });
          continue;
        }
        await writeAtomically(
          journalPath,
          JSON.stringify({ ...journal, phase: PHASE_ROLLED_BACK }),
        );
        await fs.rm(directory, { recursive: true, force: true });
        recovered += 1;
      } catch {
        const attempts = recoveryAttempts(journal) + 1;
        if (attempts < MAX_RECOVERY_ATTEMPTS) {
          await writeAtomically(
            journalPath,
            JSON.stringify({ ...journal, recovery_attempts: attempts }),
          );
          continue;
        }
        await store.markProjectState(journal.project_id, "error").catch(() => undefined);
        await fs.rm(directory, { recursive: true, force: true });
      }
    }
  }
  return recovered;
}

function tableFromRows<T extends object>(schema: Schema, rows: readonly T[]): Table {
  const columns = schema.fields.map((field) =>
    vectorFromArray(
      rows.map((row) =>
        arrowValue(field.type.constructor.name, (row as Record<string, unknown>)[field.name]),
      ),
      field.type,
    ),
  );
  const data = makeData({
    type: new Struct(schema.fields),
    length: rows.length,
    nullCount: 0,
    children: columns.map((column) => column.data[0]),
  } as never);
  return new Table(new RecordBatch(schema, data as never));
}

function arrowValue(typeName: string, value: unknown): unknown {
  if (typeName === "Int64" && value !== null && value !== undefined)
    return BigInt(value as number | bigint);
  return value;
}

function tableToBytes<T extends object>(schema: Schema, rows: readonly T[]): number {
  // IPC size is a stable upper bound for the in-flight Arrow representation.
  return tableToIPC(tableFromRows(schema, rows), "file").byteLength;
}

function normalizeRow(row: unknown, schema?: Schema): Record<string, unknown> {
  const result = { ...(row as Record<string, unknown>) };
  for (const name of ["size", "indexed_at", "start_byte", "end_byte", "start_line", "end_line"]) {
    if (typeof result[name] === "bigint") result[name] = Number(result[name]);
  }
  const vectorField = schema?.fields.find((field) => field.name === "vector");
  const float16 = vectorField?.type.children[0]?.type.precision === Precision.HALF;
  let vector: number[] | null = null;
  if (ArrayBuffer.isView(result.vector)) {
    vector = Array.from(result.vector as unknown as ArrayLike<number>);
  } else if (
    typeof result.vector === "object" &&
    result.vector !== null &&
    "toArray" in result.vector &&
    typeof result.vector.toArray === "function"
  ) {
    vector = Array.from(result.vector.toArray() as ArrayLike<number>);
  }
  if (vector !== null) result.vector = float16 ? vector.map(float16ToNumber) : vector;
  return result;
}

function float16ToNumber(value: number): number {
  const sign = value & 0x8000 ? -1 : 1;
  const exponent = (value >>> 10) & 0x1f;
  const fraction = value & 0x3ff;
  if (exponent === 0) return sign * fraction * 2 ** -24;
  if (exponent === 0x1f) return fraction === 0 ? sign * Infinity : Number.NaN;
  return sign * (1 + fraction / 0x400) * 2 ** (exponent - 15);
}

function parseJournal(value: string): StagingJournal {
  const parsed: unknown = JSON.parse(value);
  if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new Error("staging journal is not an object");
  }
  const journal = parsed as Partial<StagingJournal>;
  if (typeof journal.project_id !== "string" || typeof journal.phase !== "string") {
    throw new Error("staging journal is missing required fields");
  }
  return journal as StagingJournal;
}

function journalNumber(
  journal: StagingJournal,
  key: "files_version" | "chunks_version" | "references_version",
): number {
  const value = journal[key];
  if (!Number.isInteger(value) || (value as number) < 0)
    throw new Error(`invalid ${key} in staging journal`);
  return value as number;
}

function recoveryAttempts(journal: StagingJournal): number {
  return Number.isInteger(journal.recovery_attempts) && (journal.recovery_attempts ?? 0) >= 0
    ? (journal.recovery_attempts as number)
    : 0;
}

function addOnce(values: string[], value: string): void {
  if (!values.includes(value)) values.push(value);
}

function isMissing(error: unknown): boolean {
  return typeof error === "object" && error !== null && "code" in error && error.code === "ENOENT";
}
